Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -200,59 +205,97 @@ proof_wanted Language.isContextFree_iff {L : Language T} :
-
-lemma ContextFreeGrammar.reverse_derives (g : ContextFreeGrammar T) {s : List (Symbol T g.NT)}
-    (hgs : g.reverse.Derives [Symbol.nonterminal g.reverse.initial] s) :
-    g.Derives [Symbol.nonterminal g.initial] s.reverse := by
-  induction hgs with
-  | refl =>
-    rw [List.reverse_singleton]
-    apply ContextFreeGrammar.Derives.refl
-  | tail _ orig ih =>
-    apply ContextFreeGrammar.Derives.trans_produces ih
-    obtain ⟨r, rin, rewr⟩ := orig
-    simp only [ContextFreeGrammar.reverse, List.mem_map] at rin
-    obtain ⟨r₀, rin₀, rfl⟩ := rin
-    refine ⟨r₀, rin₀, ?_⟩
-    rw [ContextFreeRule.rewrites_iff] at rewr ⊢
-    rcases rewr with ⟨p, q, rfl, rfl⟩
-    use q.reverse, p.reverse
-    simp [ContextFreeRule.reverse]
-
-lemma ContextFreeGrammar.reverse_mem_language_of_mem_reverse_language (g : ContextFreeGrammar T)
-    {w : List T} (hgw : w ∈ g.reverse.language) :
-    w.reverse ∈ g.language := by
-  convert g.reverse_derives hgw
-  simp [List.map_reverse]
-
-lemma ContextFreeGrammar.mem_reverse_language_iff_reverse_mem_language (g : ContextFreeGrammar T)
-    (w : List T) :
-    w ∈ g.reverse.language ↔ w.reverse ∈ g.language := by
-  refine ⟨reverse_mem_language_of_mem_reverse_language _, fun hw => ?_⟩
-  rw [← ContextFreeGrammar.reverse_involutive g] at hw
-  rw [← List.reverse_reverse w]
-  exact g.reverse.reverse_mem_language_of_mem_reverse_language hw
+@[simp] lemma reverse_reverse (g : ContextFreeGrammar T) : g.reverse.reverse = g := by
+  simp [reverse, Finset.map_map]
+
+lemma reverse_involutive : Involutive (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_reverse
+
+lemma reverse_bijective : Bijective (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_involutive.bijective
+
+lemma reverse_injective : Injective (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_involutive.injective
+
+lemma reverse_surjective : Surjective (reverse : ContextFreeGrammar T → ContextFreeGrammar T) :=
+  reverse_involutive.surjective
+
+lemma produces_reverse : g.reverse.Produces u.reverse v.reverse ↔ g.Produces u v :=
+  (Equiv.ofBijective _ ContextFreeRule.reverse_bijective).exists_congr
+    (by simp [ContextFreeRule.reverse_involutive.eq_iff])
+
+alias ⟨_, Produces.reverse⟩ := produces_reverse
+
+@[simp] lemma produces_reverse_comm : g.reverse.Produces u v ↔ g.Produces u.reverse v.reverse :=
+  (Equiv.ofBijective _ ContextFreeRule.reverse_bijective).exists_congr
+    (by simp [ContextFreeRule.reverse_involutive.eq_iff])
+
+protected lemma Derives.reverse (hg : g.Derives u v) : g.reverse.Derives u.reverse v.reverse := by
+  induction hg with
+  | refl => rfl
+  | tail _ orig ih => exact ih.trans_produces orig.reverse
+
+lemma derives_reverse : g.reverse.Derives u.reverse v.reverse ↔ g.Derives u v :=
+  ⟨fun h ↦ by convert h.reverse <;> simp, .reverse⟩
+
+@[simp] lemma derivees_reverse_comm : g.reverse.Derives u v ↔ g.Derives u.reverse v.reverse := by
```

Comment:
```suggestion
@[simp] lemma derives_reverse_comm : g.reverse.Derives u v ↔ g.Derives u.reverse v.reverse := by
```

Reply:
I thought I already fixed this typo. Apparently, it was on the other branch.